jupyterhub_chameleon/handler.py

```diff
@@ -184,76 +137,116 @@ async def _fetch_new_token(self, refresh_token):
         return resp_json
 
 
-class AccessTokenHandler(AccessTokenMixin, APIHandler):
+class UserRedirectExperimentHandler(AccessTokenMixin, BaseHandler):
+    """Redirect spawn requests to user servers.
+
+    /import?{query vars} will spawn a new experiment server
+    Server will be initialized with a given artifact/repository.
+    """
+
+    artifact = None
+    version = None
+    contents_url = None
+    ownership = None
+
+    @authenticated
     async def get(self):
-        if not self.current_user:
-            raise HTTPError(401, "Authentication with API token required")
+        base_spawn_url = url_path_join(
+            self.hub.base_url, "spawn", self.current_user.name
+        )
 
         if self.request.query:
             query = dict(parse_qsl(self.request.query))
-            source = query.get("source", "unknown")
-        else:
-            source = "unknown"
+            artifact_id = query.get("artifact")
```

Comment:
This is a big if statement now. Given import doesn't make sense except when all the params are specified, refactoring to use early returns breaks up the logic in to smaller chunks.

```python
if not self.request.query:
    # malformed, import requires query params to function.
    raise Exception()

query = dict(parse_qsl(self.request.query))
artifact_id = query.get("artifact")
if not artifact_id:
    # malformed
    raise Exception()

artifact, version, contents_url = self._fetch_artifact(artifact_id, query.get("version", 0))

# do stuff with artifact, set ownership, etc.
# and fixup the spawn url
spawn_url = url_path_join(base_spawn_url, "extra stuff")
```